explainers/prioritized-task-scheduling.md

```diff
@@ -0,0 +1,856 @@
+
+ - enable prioritizing select async I/O main thread tasks, e.g. `fetch()` and async `<script>`;
+
+ - enable the browser to make better internal scheduling decisions by being aware of userland task
+   priorities;
+
+ - provide a cohesive set of scheduling APIs using modern web primitives, e.g. `AbortController`,
+   promises, etc.
+
+
+## Non-goals
+
+ - It's a non-goal to to replace every userland scheduler. Rather, our goal is to provide primitives
+   that these schedulers, and pages in general, can use to improve scheduling/performance.
+
+ - It's a non-goal to change JavaScript's [run-to-completion](https://developer.mozilla.org/en-US/docs/Web/JavaScript/EventLoop#run-to-completion)
+   semantics. Making JavaScript tasks preemptable would significantly improve responsiveness, but
+   such a paradigm shift is outside the scope of this proposal.
+
+
+## User research
+
+No user research was performed specifically for this proposal, but there have been studies on input
+latency in computing, some of which is discussed in [this event-timing
+issue](https://github.com/w3c/event-timing/issues/118).
+
+
+## Proposal
+
+### Task Priorities
+
+(See also the original [`scheduler` explainer](./prioritized-post-task.md#priorities) and
+[specification](https://wicg.github.io/scheduling-apis/#sec-task-priorities).)
+
+The proposal centers around a small set of semantic task priorities, which are used to schedule
+tasks and continuations, and which are integrated into new and existing APIs. Semantically
+meaningful naming helps developers understand when it is appropriate to use a given priority and
+enable easier coordination, and there is precedent in other systems. Similar priorities (and a
+similarly small set) can be found in other platforms like Apple's
+[QoSClass](https://developer.apple.com/documentation/dispatch/dispatchqos/qosclass) and Chromium's
+internal [browser task
+queues](https://source.chromium.org/chromium/chromium/src/+/261ad5cb51f1dbf3385af53218512796602100ed:content/browser/scheduler/browser_task_queues.h).
+
+1. **user-blocking**: tasks that block a user from interacting with and using the app. This could be
+   (chunked) work that is directly in response to user input, or updating in-viewport UI state.
+
+   User-blocking tasks are meant to have a higher priority in the event loop compared to other JS
+   tasks, i.e. they are prioritized by UA schedulers.
+
+2. **user-visible**: tasks that will be visible to the user, but either not immediately or do not
+   block the user from interacting with the page. These tasks are either less important or less
+   urgent than user-blocking tasks.
+
+   This is the default priority used for `postTask()` and the `TaskController`
+   constructor, and it is meant to be scheduled by UAs similarly to other scheduling methods, e.g.
+   same-window `postMessage` and `setTimeout(,0)`.
+
+3. **background**: Background tasks are low priority tasks that are not time-sensitive and not
```

Comment:
Out of curiosity, what would the priority of a "time-sensitive, but not visible to the user" task be? Or is there an assumption that such tasks don't exist / aren't useful?

Reply:
Anything time sensitive should use either "user-visible" at least (w/ "user-blocking" reserved for the most important things, since those tasks can lead to starvation) since "background" task are prone to starvation. These buckets might not fit all sites/apps, but the hope was they're broad for most cases. Some sites/schedulers might event create a mapping of more specific types to these (like we have in blink).

Reply:
Makes sense, thanks!